cli/ops/fs.rs

```diff
@@ -577,3 +586,73 @@ fn op_cwd(
   let path_str = path.into_os_string().into_string().unwrap();
   Ok(JsonOp::Sync(json!(path_str)))
 }
+
+#[allow(dead_code)]
+struct FsWatcher {
+  watcher: RecommendedWatcher,
+  receiver:
+    crossbeam_channel::Receiver<Result<notify::event::Event, notify::Error>>,
+}
+
+impl Resource for FsWatcher {}
+
+#[derive(Deserialize)]
+struct OpenWatcherArgs {
+  recursive: bool,
+  paths: Vec<String>,
+  debounce: u64,
+}
+
+pub fn op_watch(
+  state: &ThreadSafeState,
+  args: Value,
+  _zero_copy: Option<PinnedBuf>,
+) -> Result<JsonOp, ErrBox> {
+  let args: OpenWatcherArgs = serde_json::from_value(args)?;
+  let (tx, rx) = crossbeam_channel::unbounded();
+  let mut watcher = watcher(tx, Duration::from_millis(args.debounce))?;
+
+  let recursive_mode: RecursiveMode = if args.recursive {
+    RecursiveMode::Recursive
+  } else {
+    RecursiveMode::NonRecursive
+  };
+  for path in &args.paths {
+    state.check_read(path)?;
+    watcher.watch(path, recursive_mode)?;
+  }
+
+  let watcher_resource = FsWatcher {
+    watcher,
+    receiver: rx,
+  };
+  let mut table = state.lock_resource_table();
+  let rid = table.add("fsWatcher", Box::new(watcher_resource));
+  Ok(JsonOp::Sync(json!(rid)))
+}
+
+#[derive(Deserialize)]
+struct PollWatcherArgs {
+  rid: i32,
+}
+
+pub fn op_fs_poll_watcher(
+  state: &ThreadSafeState,
+  args: Value,
+  _zero_copy: Option<PinnedBuf>,
+) -> Result<JsonOp, ErrBox> {
+  let args: PollWatcherArgs = serde_json::from_value(args)?;
+  let state = state.clone();
+
+  blocking_json(false, move || {
+    let mut table = state.lock_resource_table();
+    let resource = table
+      .get_mut::<FsWatcher>(args.rid as u32)
+      .ok_or_else(bad_resource)?;
+    // blocks
+    let result = resource.receiver.recv().map_err(ErrBox::from)?;
+    let event = result.map_err(ErrBox::from)?;
+    let serialized = serde_json::to_string(&event).expect("Foobar");
+    Ok(json!(serialized))
```

Comment:
Note to self - this is bad, it obviously blocks while holding a mutex. This must be rewritten into a `Future` using probably `try_recv` but then a waker is needed - to be figured